docs/src/developer/conventions.md

```diff
@@ -0,0 +1,173 @@
+```@meta
+CurrentModule = Nemo
+```
+
+# Conventions
+
+AbstractAlgebra and Nemo have adopted a number of conventions to help maintain
+a uniform codebase.
+
+## Code conventions
+
+### Function and type names
+
+Names of types in Julia follow the convention of CamelCase where the first
+letter of each word is capitalised, e.g. `Int64` and `AbstractString`.
+
+Function/method names in Julia use all lowercase with underscores between
+the words, e.g. `zip` and `jacobi_symbol`.
+
+We follow these conventions in Nemo with some exceptions:
+
+* When interfacing C libraries the types use the same spelling and
+capitalisation in Nemo as they do in C, e.g. the Flint library's `fmpz_poly`
+remains uncapitalised in Nemo.
+
+* Types such as `gfp_poly` which don't exist under that name on the C side
+also use the lowercase convention as they wrap an actual C type which must be
+split into more than one type on the Julia side. For example `nmod_poly` and
+`gfp_poly` on the Julia side both represent Flint `nmod_poly`'s on the C side.
+
+* Types of rings and fields, modules, maps, etc. are capitalised whether they
+correspond to a C type or not, e.g. `FqNmodFiniteField` for the type of an
+object representing the field that `fq_nmod`'s belong to.
+.
+* We omit an underscore if the first word of a method is "is" or "has", e.g.
+`iseven`.
+
+* Underscores are omitted if the method name is already well established
+without an underscore in Julia itself, e.g. `setindex`.
+
+* Inner and outer constructors with the same name as a type use the same
+spelling and capitalisation as that type, e.g. `fmpz(1)`.
+
+* Functions for creating rings, fields, modules, maps, etc. (rather than the
+elements thereof) use CamelCase, e.g. `PolynomialRing`. We refer to these 
+functions as parent constructors.
+
+* We prefer words to not be abbreviated, e.g. `denominator` instead of `den`.
+
+* Exceptions always exist where the result would be offensive in any major
+spoken language (example omitted).
+
+It is easy to find counterexamples to virtually all these rules. However we
+have been making efforts to remove the most egregious cases from our codebase
+over time. As perfect consistency is not possible, work on this has to at
+times take a back seat.
+
+### Spacing and tabs
+
+All function bodies and control blocks should be indented using spaces.
+
+A survey of existing code shows 2, 3 or 4 space indenting commonly used in our
+files. Values outside this range should not be used.
+
+When contributing to an existing file, follow the majority convention in that
+file. Consistency within a file is valued highly.
+
+If you are new to Nemo development and do not already have a very strong
+preference, new files should be started with 3 space indenting. This maximises
+the likelihood that copy and paste between files will be straighforward, though
+modern editors ease this to some degree.
+
+Prototypes in docstrings should have four spaces before them.
```

Comment:
```suggestion
Signatures in docstrings should have four spaces before them.
```

Reply:
There's no difference in my opinion. A prototype is a function signature without a function body, which the line in the docs certainly qualifies as. Nonetheless I changed the word anyway, since it makes no difference to me.

Reply:
There are no function prototypes on julia.

Reply:
Not quite true. Julia provides the @cfunction macro for prototyping C functions (the documentation even uses this word to describe it).

As far as C is concerned, Kernighan and Ritchie defined a prototype as a compiler directive (e.g. a forward declaration). But many other languages have function prototypes as well and the meaning varies a bit from K&R's definition.

I accept that the main definition appears to be something that directs the compiler with regard to parameter types, function name and return type. E.g. a line in a .h file describing a function is properly referred to as a prototype.

I'll accept that signature is a better term for exactly the same thing when it appears in human readable documentation. But for example, the @doc statement that corresponds to the signature I'm talking about above is certainly a prototype. It seems like a distinction without too much substance to me, like compiler vs transpiler.

Reply:
Also from Wikipedia, "A function signature consists of the function prototype. It specifies the general information about a function like the name, scope and parameters."

Reply:
The basic gist is that a signature is a more general thing than a function signature. Other things can have signatures, e.g. type signature, method signature, etc. Together they are usually referred to as type signatures. A function prototype is the type signature of a function (along with the function name, possibly depending whether the language is nominative or not).

This is not the definition as given by K&R for sure. But that's a pretty narrow definition.

Reply:
Still other (equally authoritative CS resources) say that a signature does not include the return type, whereas a prototype includes name, return type and parameter type. So it's clear there are many competing definitions.